CDAP-13924 Add Capability Annotations APIs

rohitsinha54 opened the pull request:
[Issue](https://issues.cask.co/browse/CDAP-13924)
[Design](https://wiki.cask.co/display/CE/Improving+Plugins+User+Experience+Across+Different+Modes)

Introduces ```@ Capability``` annotation which will be used to define capability of a plugin plugin filtering.

For details please review [API section](https://wiki.cask.co/display/CE/Improving+Plugins+User+Experience+Across+Different+Modes#ImprovingPluginsUserExperienceAcrossDifferentModes-API) of the design.
 
Build: https://builds.cask.co/browse/CDAP-DUT6579-1

rohitsinha54 committed:
CDAP-13924 Add Capability Annotations APIs

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java:
```diff
@@ -0,0 +1,24 @@
+package co.cask.cdap.api.annotation;
+
+/**
+ * Defines the different capability options for {@link Plugin} to specify.
```
link to the Capability annotation, not to Plugin.

rohitsinha54 replied:
Done.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java:
```diff
@@ -0,0 +1,64 @@
+import java.lang.annotation.Target;
+
+/**
+ * <p>Annotates capabilities of a plugin.</p>
```
It's easier to read in code with just newlines instead of paragraph tags. Does it render differently If you just have newlines?

rohitsinha54 replied:
Yes the newlines are not retained in rendering if we don't use the tags. I added new line for readability and kept the tags.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java:
```diff
@@ -0,0 +1,64 @@
+ * </li>
+ * <pre>
+ *      {@literal @}Plugin(type = BatchSource.PLUGIN_TYPE)
+ *      {@literal @}Name("PII")
```
Lets name it something different in case people get confused.

rohitsinha54 replied:
Done.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java:
```diff
@@ -0,0 +1,64 @@
+ * <p>Pre-defined system capabilities are defined in {@link Capabilities}.</p>
+ * <p>Plugins can also be annotated with custom value to specify capability with business rules.
+ * The capability value is case insensitive.</p>
+ * <p>If a plugin does not specify any capability or the {@link #value()} array is empty then the default capability
```
thinking about this some more, it means a plugin annotated with just 'pii' can't run anywhere. It might be better to define it such that if no system defined capability is specified, it gets all system defined capabilities. So if a plugin is just tagged with 'pii', it can be run natively or remotely.

rohitsinha54 replied:
That was the idea only if system defined capability is present the it overrides the default. The presence or absence of user defined capability does not affect override system defined. I will add that in documentation.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java:
```diff
@@ -0,0 +1,64 @@
+ * <p>According to default capability the plugin will be considered as capable of all system defined
+ * {@link Capabilities} but not with any user defined capability options.
+ * Custom capabilities needs to be specifically defined.</p>
+ * <p>Usage Examples:</p>
```
It would be good to have an example where no Capability is specified.

rohitsinha54 replied:
Done added more usage examples.

albertshau reviewed (commented):
Are those links in the commit message or just in the PR message? I would avoid putting links in commit messages, as they can change or be removed.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java:
```diff
@@ -0,0 +1,27 @@
+  public static final String NATIVE = "native"; // defines capability of running in native
+  public static final String REMOTE = "remote"; // defines capability of running remotely
+
+  // spark capabilities are purposefuly not defined as we want to disallow usage of spark 1 or specifying capability
```
Seems irrelevant to have this comment.

rohitsinha54 replied:
Done.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java:
```diff
@@ -0,0 +1,27 @@
+ * Defines the different capability options for {@link Plugin} to specify.
+ */
+public final class Capabilities {
+  public static final String NATIVE = "native"; // defines capability of running in native
```
The comment should be a javadoc for the public constant.

rohitsinha54 replied:
Done.

rohitsinha54 commented:
@albertshau : The link is only in PR description. 
@albertshau @chtyim: Thanks for the prompt review. Addressed all comments. Please have another look.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java:
```diff
@@ -16,12 +16,25 @@
+ * <p>Defines the different capability options for {@link Capability} to specify.
+ * These are system defined capabilities.</p>
+ *
+ * <p>If {@link Capability} does not specifically define any of the capability from {@link Capabilities} then
```
Is this true? Meaning if a plugin is annotated with `@Capability("xyz")` then it will implicitly have `native` and `remote`?

chtyim replied:
Or should the statement be `If a {@link Plugin} does not define any {@link Capability}, then the {@link Capability#default} system capability will be assumed`?

albertshau replied:
if a plugin only annotated with 'xyz' does not get any system capabilities, then it means it can't actually be used anywhere, which seems weird. 

maybe we should differentiate between system and custom capabilities? If no system capability is given, it gets all system capabilities.

chtyim replied:
If the set of system capabilities is keep expanding, it might be hard for plugin to catch up with.

albertshau replied:
I think the difficulty comes from the fact that we're trying to use the same annotation for two separate use cases. For system things, we really want the plugin to specify what it requires, not what it's capable of. When we run, everything a plugin requires must be supported by the provisioner. 

And for custom things, it's the opposite. We want to specify what capabilities it has, so that an admin can configure CDAP to require a set of capabilities.

What if we separate these things? Provisioners are responsible for specifying what capabilities they have. Native would have {transactional}, dataproc would have {gcp}, emr would have {aws}. Transactional plugins would be annotated as transactions. Plugins that work anywhere would not need any annotations.

For the pii case, custom plugins would annotate that they have a 'pii' capability.

At a CDAP level, you can configure a whitelist of plugin capabilities. If you want only pii plugins in CDAP, configure 'plugin.capability.whitelist' to 'pii'.
You can also configure a blacklist of plugin requirements. If you don't want transactional plugins to appear, configure 'plugin.requirement.blacklist' to 'transactional'.

rohitsinha54 replied:
Yes, if a plugin does not specify any capability from System `Capabilities` it will be considered to be capable of all system defined capabilities. This applies even if they only specify custom capabilities.

rohitsinha54 replied:
```
If the set of system capabilities is keep expanding, it might be hard for plugin to catch up with.
```

In this case the plugin will specifically specify what they are capable once (when they are no more capable of all system defined capabilities). Once it is specified as long as now new capability is added to the plugin the specification does not need to be changed. If new capability is added to the plugin then we can expect the plugin developer to a update the specification too.

rohitsinha54 replied:
We had in person discussion and decided that we will separate out the annotation for system capabilities and user capabilities. I will update the design with this new decision and update the code here.

rohitsinha54 replied:
https://wiki.cask.co/display/CE/Improving+Plugins+User+Experience+Across+Different+Modes?focusedCommentId=8716406#comment-8716406

rohitsinha54 replied:
@chtyim @albertshau  Update on design: https://wiki.cask.co/display/CE/Improving+Plugins+User+Experience+Across+Different+Modes?focusedCommentId=8716406#comment-8716406

Please have a look. I will update the code with this shortly.

sreevatsanraman commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java:
```diff
@@ -0,0 +1,40 @@
+package co.cask.cdap.api.annotation;
+
+/**
+ * <p>Defines the different capability options for {@link Capability} to specify.
```
Capability options for capability is recursive. Can you elaborate on what it means.

anwar6953 commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java:
```diff
@@ -0,0 +1,40 @@
+ * <p>According to default system capability a plugin is considered to be capable of all the capabilities defined in
+ * {@link Capabilities}.</p>
+ */
+public final class Capabilities {
```
Are the values in here parallel to the values in `ClusterMode`?

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java:
```diff
@@ -0,0 +1,71 @@
+ *
+ * <p>Usage Examples:</p>
+ * <ul>
+ * <li><b>Omitting Requirements:</b> A plugin can chose to not specify any requirement by not specifying a
```
chose -> choose

rohitsinha54 replied:
👍

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java:
```diff
@@ -0,0 +1,71 @@
+ * <li><b>Omitting Requirements:</b> A plugin can chose to not specify any requirement by not specifying a
+ * {@link Requirements} annotation. In this case the plugin will considered to be capable or run everywhere.</li>
+ * <pre>
+ *     {@literal @}Plugin(type = BatchSource.PLUGIN_TYPE)
```
seems like the `@` is off by a single space. Or does it get rendered differently?

rohitsinha54 replied:
It gets rendered differently. the spacing is to render it correctly. 
![image](https://user-images.githubusercontent.com/1472867/44240985-f0eb5400-a175-11e8-8d9b-dfaf820b76e9.png)

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java:
```diff
@@ -0,0 +1,71 @@
+ * <pre>
+ *     {@literal @}Plugin(type = BatchSource.PLUGIN_TYPE)
+ *     {@literal @}Name("TransactionalSource")
+ *     {@literal @}Capability({"transactions"})
```
Capability -> Requirements

rohitsinha54 replied:
Done.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java:
```diff
@@ -0,0 +1,71 @@
+ *      }
+ *   </pre>
+ * <li><b>Specifying multiple requirements:</b> A plugin can also specify multiple requirements. For example if a
+ * plugin needs 'transactions' and 'kms' to run this can specified as below.</li>
```
lets use 'securestore' instead of 'kms' to be consistent with terminology.

rohitsinha54 replied:
Done.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java:
```diff
@@ -0,0 +1,71 @@
+ *
+ * <p>Requirements are case insensitive.</p>
+ *
+ * <p>If a plugin does not specify any {@link Requirements} or the {@link #value()} array is empty then it assumed
```
`If a plugin does not annotated with {@link Requirements} or the annotation {@link #value()} is empty, then it is assumed that the plugin..`

rohitsinha54 replied:
Done.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java:
```diff
@@ -0,0 +1,71 @@
+ * {@link Requirements} annotation. In this case the plugin will considered to be capable or run everywhere.</li>
+ * <pre>
+ *     {@literal @}Plugin(type = BatchSource.PLUGIN_TYPE)
+ *     {@literal @}Name("NoRequirementSource")
```
This is not a good naming example for a plugin. Probably use `CSVParser`

rohitsinha54 replied:
Done.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java:
```diff
@@ -0,0 +1,71 @@
+ * <pre>
+ *     {@literal @}Plugin(type = BatchSource.PLUGIN_TYPE)
+ *     {@literal @}Name("TransactionalSource")
+ *     {@literal @}Capability({"transactions"})
```
For single value, you don't need the `{}`.

rohitsinha54 replied:
Done.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java:
```diff
@@ -0,0 +1,71 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface Requirements {
+  String[] value() default {};
```
Since requirements are system specific, shouldn't we use Enum instead of strings? Also we should define those system abilities here.

rohitsinha54 replied:
We decided provisioner can have their own capabilities defined like 'transactional', 'gcp', 'aws', 'azure', 'securestore', 's3' etc etc and we do not want to stick to a defined requirements model.

albertshau replied:
Some are system specific, but some are things we wanted to discourage (like spark version), and others are defined by the provisioner. Maybe we should include the known system ones in a constants class like earlier?

chtyim replied:
but the platform needs to have a contract on system capabilities, such as `transaction`, right? How does the provisioner implementations know about that? Maybe those constants should be defined here?

rohitsinha54 replied:
Done.

albertshau reviewed (approved):
lgtm

rohitsinha54 commented:
Waiting for build: https://builds.cask.co/browse/CDAP-DUT6579-4

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java:
```diff
@@ -0,0 +1,71 @@
+ * <pre>
+ *     {@literal @}Plugin(type = BatchSource.PLUGIN_TYPE)
+ *     {@literal @}Name("SecureTable")
+ *     {@literal @}Requirements({"transactions", "securestore"})
```
The "securestore" may not be a good example as it is not decided by the Plugin. Any marco that is `$secure{...}` will lookup secure store by the data pipeline.

rohitsinha54 replied:
Done.

rohitsinha54 commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java:
```diff
@@ -0,0 +1,26 @@
+  /**
+   * Defines transactional capabilities
+   */
+  public static final String TRANSACTIONS = "transactions";
```
More will be added as and when needed down the line, during implementation

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * <p>Defines the different capability options for a {@link Plugin} to specify it's {@link Requirements}.</p>
+ */
+public final class Capabilities {
```
No need to create a new class for this. You can define constants in the `Requirements` annotation.

rohitsinha54 replied:
Awesome. Thanks.

chtyim reviewed (approved):
LGTM
